README.md
# pdaj2

CHANGELOG
export_beam_integrals 1.0.1 (April 11, 2016)

    * Vagrant: Added the vbguest plugin as a dependency, to keep VirtualBox
      Guest Additions up to date.

    * Vagrantfile: The configuration `docker.version' has been deprecated.

export_beam_integrals 1.0.0 (November 17, 2015)

    * Initial release.


double_pendulum/tasks/worker.py
from math import log

#from beam_integrals.beam_types import BaseBeamType 
#from beam_integrals.exceptions import UnableToGuessScaleFunctionError
#from beam_integrals.integrals import BaseIntegral, integrate

from ..app import app


SCALE_FACTOR_HELPER = 2.

import sys
import numpy as np
from scipy.integrate import odeint

#based on https://scipython.com/blog/the-double-pendulum/

# The gravitational acceleration (m.s-2).
g = 9.81


def deriv(y, t, L1, L2, m1, m2):
    """Return the first derivatives of y = theta1, z1, theta2, z2."""
    theta1, z1, theta2, z2 = y

    c, s = np.cos(theta1-theta2), np.sin(theta1-theta2)

    theta1dot = z1
    z1dot = (m2*g*np.sin(theta2)*c - m2*s*(L1*z1**2*c + L2*z2**2) -
             (m1+m2)*g*np.sin(theta1)) / L1 / (m1 + m2*s**2)
    theta2dot = z2
    z2dot = ((m1+m2)*(L1*z1**2*s - g*np.sin(theta2) + g*np.sin(theta1)*c) +
             m2*L2*z2**2*s*c) / L2 / (m1 + m2*s**2)
    return theta1dot, z1dot, theta2dot, z2dot

def solve(L1, L2, m1, m2, tmax, dt, y0):
    t = np.arange(0, tmax+dt, dt)

    # Do the numerical integration of the equations of motion
    y = odeint(deriv, y0, t, args=(L1, L2, m1, m2))
    theta1, theta2 = y[:,0], y[:,2]

    # Convert to Cartesian coordinates of the two bob positions.
    x1 = L1 * np.sin(theta1)
    y1 = -L1 * np.cos(theta1)
    x2 = x1 + L2 * np.sin(theta2)
    y2 = y1 - L2 * np.cos(theta2)

	return theta1, theta2, x1, y1, x2, y2


@app.task
def simulate_pendulum_instance(L1, L2, m1, m2, tmax, dt, theta1_init, theta2_init):
	 
	y0 = np.array([
		theta1_init,
		0.0,
		theta2_init,
		0.0
	])
    #vracam i thete da znam za cega su ovi rezultati (da ih posle mogu
    #pisati u fajl)
	return theta1_init, theta2_init, solve(L1, L2, m1, m2, tmax, dt, y0)

'''
@app.task
def compute_integral(integral_id, beam_type_id, m, t, v, n):
    beam_type = BaseBeamType.coerce(beam_type_id)
    integral = BaseIntegral.coerce(integral_id)
    max_mode = app.conf.BEAM_INTEGRALS_MAX_MODE
    decimal_precision = app.conf.BEAM_INTEGRALS_DECIMAL_PRECISION
    normalize_integrals_smaller_than = app.conf.BEAM_INTEGRALS_NORMALIZE_INTEGRALS_SMALLER_THAN

    result, error = integrate(
        integral, beam_type,
        a=1.,
        m=m, t=t, v=v, n=n,
        decimal_precision=decimal_precision,
        error=True
    )

    # Normalize the `result` to zero for small values, as per the section 3.4 of my PhD thesis
    if abs(result) <= normalize_integrals_smaller_than:
        result = 0.
        scale_factor = 0
    else:
        try:
            scale_factor = integral.guess_scale_factor(beam_type, m, t, v, n)
        except UnableToGuessScaleFunctionError:
            scaled_result = integrate(
                integral, beam_type,
                a=SCALE_FACTOR_HELPER,
                m=m, t=t, v=v, n=n,
                decimal_precision=decimal_precision
            )
            scale_factor = round(log(scaled_result / result, SCALE_FACTOR_HELPER))

    cache_key = integral.cache_key(m, t, v, n, max_mode=max_mode)
    data = {
        'integral_float64': float(result),
        'error_float64': float(error),
        'scale_factor': int(scale_factor),
        'integral_str': str(result),
        'error_str': str(error),
    }
    return cache_key, data
'''

@app.task
def combine_computed_integrals_into_a_table(computed_integrals, integral_id):
    return integral_id, dict(computed_integrals)


double_pendulum/docker/docker-entrypoint.sh
#!/bin/bash
set -e

record_experiment_status() {
    if [ "$COMPUTER_TYPE" = 'server' ]; then # Record the experiment status only when running on the server
        STATUS_DIR="${RESULTS_DIR}/status"
        mkdir -p "$STATUS_DIR"

        STATUS_FILE="${STATUS_DIR}/$1"
        if [ ! -f "$STATUS_FILE" ]; then
            # Current local date and time, in ISO 8601 format (microseconds and TZ removed)
            echo `date +%FT%T` > $STATUS_FILE
        fi
    fi
}

if [ "$1" = 'supervisord' ]; then # Normal startup, i.e. no explicit command passed to 'docker run'
    record_experiment_status 'init'

    if [ -n "$DOCKER_CONFIG_SCRIPT" ]; then
        SCRIPT_PATH="/$DOCKER_CONFIG_SCRIPT"
        echo "Running the configuration script $SCRIPT_PATH..."
        bash -c "$SCRIPT_PATH"
    fi

    record_experiment_status 'configured'

    exec "$@"
fi

exec "$@"


double_pendulum/double_pendulum/celeryconfig.py
from multiprocessing import cpu_count
import os

from kombu import Queue


## Environment based settings

MAX_CPU_CORES = os.getenv('MAX_CPU_CORES', cpu_count())
SERVER_NAME = os.getenv('SERVER_NAME', 'localhost')
AM_I_SERVER = (os.getenv('COMPUTER_TYPE') == 'server')

TMAX = int(os.getenv('TMAX', 30)) #ovo 30 je defaultna vrednost
DT  = float(os.getenv('DT', 0.01))
THETA_RESOLUTION  = int(os.getenv('THETA_RESOLUTION', 5))

if AM_I_SERVER:
    MONITORING_IS_ACTIVE = bool(int(os.getenv('MONITORING_IS_ACTIVE', '0')))
    if MONITORING_IS_ACTIVE:
        MONITORING_SERVER_NAME = os.getenv('MONITORING_SERVER_NAME', 'localhost')
        MONITORING_SERVER_PORT = int(os.getenv('MONITORING_SERVER_PORT', 2003))
        MONITORING_INTERVAL = int(os.getenv('MONITORING_INTERVAL', 30))
        MONITORING_METRIC_PREFIX = os.getenv('MONITORING_METRIC_PREFIX', 'experiments.double_pendulum')

    HDF5_COMPLIB = os.getenv('HDF5_COMPLIB', 'zlib')
    HDF5_COMPLEVEL = int(os.getenv('HDF5_COMPLEVEL', 1))

    RESULTS_DIR = os.getenv('RESULTS_DIR', '/tmp/results')
    STATUS_DIR = os.path.join(RESULTS_DIR, 'status')


## Concurrency settings

CELERYD_CONCURRENCY = MAX_CPU_CORES

# This ensures that each worker will only take one task at a time, when combined
# with late acks. This is the recommended configuration for long-running tasks.
# References:
#   * http://celery.readthedocs.org/en/latest/userguide/optimizing.html#prefetch-limits
#   * http://celery.readthedocs.org/en/latest/userguide/optimizing.html#reserve-one-task-at-a-time
#   * http://celery.readthedocs.org/en/latest/configuration.html#celeryd-prefetch-multiplier
#   * http://stackoverflow.com/questions/16040039/understanding-celery-task-prefetching
#   * https://bugs.launchpad.net/openquake-old/+bug/1092050
#   * https://wiredcraft.com/blog/3-gotchas-for-celery/
#   * http://www.lshift.net/blog/2015/04/30/making-celery-play-nice-with-rabbitmq-and-bigwig/
CELERYD_PREFETCH_MULTIPLIER = 1


## Task result backend settings

CELERY_RESULT_BACKEND = "redis://%s" % SERVER_NAME #tu cuva medjurezultate


## Message Routing

CELERY_DEFAULT_QUEUE = 'worker'
CELERY_DEFAULT_EXCHANGE = 'tasks'
CELERY_DEFAULT_ROUTING_KEY = 'worker'

if AM_I_SERVER:
    CELERY_QUEUES = (
        Queue('server',  routing_key='server'),
    )
else:
    CELERY_QUEUES = (
        Queue('worker', routing_key='worker'),
    )

class ServerTasksRouter(object):
    def route_for_task(self, task, args=None, kwargs=None):
        if task.startswith('double_pendulum.tasks.server.'):
            return {'queue': 'server'}
        
        return None

CELERY_ROUTES = (
    ServerTasksRouter(),
)


## Broker Settings

BROKER_URL = "amqp://%s" % SERVER_NAME
CELERY_ACCEPT_CONTENT = ['pickle', 'json']


## Task execution settings

CELERY_MESSAGE_COMPRESSION = 'bzip2'
CELERY_TASK_RESULT_EXPIRES = None
CELERY_DISABLE_RATE_LIMITS = True
CELERY_TRACK_STARTED = True		#nesto za monitoring

# This ensures that the worker acks the task *after* it's completed.
# If the worker crashes or gets killed mid-execution, the task will be returned
# to the broker and restarted on another worker.
# References:
#   * https://wiredcraft.com/blog/3-gotchas-for-celery/
#   * http://celery.readthedocs.org/en/latest/configuration.html#celery-acks-late
#   * http://celery.readthedocs.org/en/latest/faq.html#faq-acks-late-vs-retry
CELERY_ACKS_LATE = True


## Worker settings

if AM_I_SERVER:
    CELERY_IMPORTS = ['double_pendulum.tasks.server']
else:
    CELERY_IMPORTS = ['double_pendulum.tasks.worker']

# HACK: Prevents weird SymPy related memory leaks
CELERYD_MAX_TASKS_PER_CHILD = 100


## Periodic Task Server (celery beat)

if AM_I_SERVER and MONITORING_IS_ACTIVE:
    CELERYBEAT_SCHEDULE = {
        'monitor-queues': {
            'task': 'double_pendulum.tasks.server.monitor_queues',
            'schedule': MONITORING_INTERVAL,
        },
    }


double_pendulum/double_pendulum/tasks/server.py
from datetime import datetime
import os
import socket
import subprocess
import time

#import beam_integrals as bi
#from beam_integrals.beam_types import BaseBeamType
#from beam_integrals.integrals import BaseIntegral
from celery import chain, chord
from celery.exceptions import Reject
import numpy as np
import tables as tb

import double_pendulum as ebi
from ..app import app
from .worker import compute_integral, combine_computed_integrals_into_a_table


## Monitoring tasks

@app.task
def monitor_queues(ignore_result=True):
    server_name = app.conf.MONITORING_SERVER_NAME
    server_port = app.conf.MONITORING_SERVER_PORT
    metric_prefix = app.conf.MONITORING_METRIC_PREFIX

    queues_to_monitor = ('server', 'worker')
    
    output = subprocess.check_output('rabbitmqctl -q list_queues name messages consumers', shell=True)
    lines = (line.split() for line in output.splitlines())
    data = ((queue, int(tasks), int(consumers)) for queue, tasks, consumers in lines if queue in queues_to_monitor)

    timestamp = int(time.time())
    metrics = []
    for queue, tasks, consumers in data:
        metric_base_name = "%s.queue.%s." % (metric_prefix, queue)

        metrics.append("%s %d %d\n" % (metric_base_name + 'tasks', tasks, timestamp))
        metrics.append("%s %d %d\n" % (metric_base_name + 'consumers', consumers, timestamp))

    sock = socket.create_connection((server_name, server_port), timeout=10)
    sock.sendall(''.join(metrics))
    sock.close()


## Recording the experiment status
#ako je nesto vec racunao, on radi neke optimizacije
def get_experiment_status_filename(status):
    return os.path.join(app.conf.STATUS_DIR, status)

def get_experiment_status_time():
    """Get the current local date and time, in ISO 8601 format (microseconds and TZ removed)"""
    return datetime.now().replace(microsecond=0).isoformat()

@app.task
def record_experiment_status(status):
    with open(get_experiment_status_filename(status), 'w') as fp:
        fp.write(get_experiment_status_time() + '\n')


## Seeding the computations

def parametric_sweep(theta_resolution, tmax, dt):
    # Pendulum rod lengths (m), bob masses (kg).
    L1, L2 = 1.0, 1.0
    m1, m2 = 1.0, 1.0

    # Maximum time, time point spacings (all in s).
    #tmax, dt = 30.0, 0.01

    theta1_inits = np.linspace(0, 2*np.pi, theta_resolution)
    theta2_inits = np.linspace(0, 2*np.pi, theta_resolution)

    import itertools
    t1t2_inits = itertools.product(theta1_inits, theta2_inits)
	return ((L1, L2, m1, m2, tmax, dt, t1t2_i[0], t1t2_i[1]) for t1t2_i in t1t2_inits)

@app.task
def seed_computations(ignore_result=True):
    if os.path.exists(get_experiment_status_filename('started')):
        raise Reject('Computations have already been seeded!')

    record_experiment_status.si('started').delay()
    
    _tmax = app.conf.TMAX
    _theta_res = app.conf.THETA_RESOLUTION
    _dt = app.conf.DT

    chord(
        (
			simulate_pendulum_instance(L1, L2, m1, m2, tmax, dt, theta1_init, theta2_init)
			for (L1, L2, m1, m2, tmax, dt, theta1_init, theta2_init) in
			parametric_sweep(theta_resolution, tmax, dt)
		),
        store_computed_integral_tables.s()
    ).delay()

def double_pendulum(beam_type_id):
    # OPTIMIZATION: Compute only the canonical integrals, as per the section 3.3.1 of my PhD thesis
    canonical_integrals = (integral for integral in BaseIntegral.plugins.instances_sorted_by_id if not integral.has_parent())

    return chord(
        (compute_integral_table(integral.id, beam_type_id) for integral in canonical_integrals),
        store_computed_integral_tables.s(beam_type_id)
    )

def compute_integral_table(integral_id, beam_type_id):
    integral = BaseIntegral.coerce(integral_id)
    max_mode = app.conf.BEAM_INTEGRALS_MAX_MODE

    # OPTIMIZATION: Don't compute integrals if their modes are equivalent, as per the section 3.3.2 of my PhD thesis
    cache_keys_seen = set()
    unique_integration_variables = []
    for variables in integral.iterate_over_used_variables(max_mode=max_mode):
        cache_key = integral.cache_key(*variables, max_mode=max_mode)
        if cache_key not in cache_keys_seen: # Cache miss
            cache_keys_seen.add(cache_key)
            unique_integration_variables.append(variables)

    return chord(
        (compute_integral.s(integral.id, beam_type_id, *variables) for variables in unique_integration_variables),
        combine_computed_integrals_into_a_table.s(integral_id)
    )


## Storing the computed integral tables

def get_hdf5_table_description(used_variables, decimal_precision):
    columns = dict(
        (var, tb.UInt8Col(pos=idx))
        for idx, var in enumerate(used_variables)
    )
    
    data_start_pos = len(used_variables)
    columns['integral_float64'] = tb.Float64Col(pos=data_start_pos)
    columns['error_float64'] = tb.Float64Col(pos=data_start_pos+1)

    columns['scale_factor'] = tb.Int8Col(pos=data_start_pos+2)
    
    max_len = decimal_precision + 10 # Account for decimal dot and exponent info
    columns['integral_str'] = tb.StringCol(itemsize=max_len, pos=data_start_pos+3)
    columns['error_str'] = tb.StringCol(itemsize=max_len, pos=data_start_pos+4)

    return columns

@app.task
def store_computed_integral_tables(results):
	with open('blah.csv', 'w') as out:
		out.write(results)

'''
    beam_type = BaseBeamType.coerce(beam_type_id)
    max_mode = app.conf.BEAM_INTEGRALS_MAX_MODE
    decimal_precision = app.conf.BEAM_INTEGRALS_DECIMAL_PRECISION
    hdf5_complib = app.conf.HDF5_COMPLIB
    hdf5_complevel = app.conf.HDF5_COMPLEVEL
    results_dir = app.conf.RESULTS_DIR

    # Celery `chord` sent us a list, convert it to a `dict` for easier use
    integral_tables = dict(integral_tables)

    # These filters will be applied to all the datasets created immediately under the root group:
    #   * 'complib': Specifies the compression library to be used. Although PyTables
    #     supports many interesting compression libraries, HDF5 itself provides
    #     only 2 pre-defined filters for compression: ZLIB and SZIP. We can't
    #     use SZIP due to licensing issues, therefore ZLIB has been chosen by default as
    #     it's supported by all major HDF5 viewers (HDFView, HDF Compass, ViTables,
    #     HDF Explorer).
    #   * 'complevel': Specifies a compression level for data. Using the lowest
    #     level (1) by default, per PyTables optimization recommendations (see references).
    #   * 'shuffle': Enable the Shuffle filter to improve the compression ratio.
    #   * 'fletcher32': Enable the Fletcher32 filter to add a checksum on each
    #     data chunk.
    #
    # References:
    #   * https://www.hdfgroup.org/services/filters.html
    #   * https://www.hdfgroup.org/hdf5-quest.html#gcomp
    #   * https://www.hdfgroup.org/HDF5/faq/compression.html
    #   * http://www.pytables.org/usersguide/libref/helper_classes.html#the-filters-class
    #   * http://www.pytables.org/usersguide/optimization.html#compression-issues
    #   * http://www.pytables.org/usersguide/optimization.html#shuffling-or-how-to-make-the-compression-process-more-effective
    filters = tb.Filters(complib=hdf5_complib, complevel=hdf5_complevel, shuffle=True, fletcher32=True)

    output_filename = os.path.join(results_dir, "%s.hdf5" % beam_type.filename)
    with tb.open_file(output_filename, 'w', filters=filters) as out:
        # Add the root group metadata, as per the section 3.2 of my PhD thesis
        out.root._v_attrs.created_at = get_experiment_status_time()
        out.root._v_attrs.generator_name = 'double_pendulum'
        out.root._v_attrs.generator_version = ebi.__version__
        out.root._v_attrs.beam_integrals_version = bi.__version__
        out.root._v_attrs.decimal_precision = decimal_precision
        out.root._v_attrs.max_mode = max_mode
        out.root._v_attrs.beam_type_name = beam_type.name
        out.root._v_attrs.beam_type_id = beam_type.id

        for integral in BaseIntegral.plugins.instances_sorted_by_id:
            if integral.has_parent(): 
                # OPTIMIZATION: This integral is the same as his parent, just create a hard link,
                # as per the section 3.3.1 of my PhD thesis
                parent = BaseIntegral.plugins.id_to_instance[integral.parent_id()]
                out.create_hard_link(
                    where='/',
                    name=integral.name,
                    target='/' + parent.name
                )

                # Create a hard link to the parent's index as well
                out.create_hard_link(
                    where='/',
                    name='_i_' + integral.name,
                    target='/_i_' + parent.name
                )

                continue # No further processing needed, skip to the next integral

            # Help PyTables determine the optimal chunk size
            num_rows = max_mode ** len(integral.used_variables)

            table = out.create_table(
                where='/',
                name=integral.name,
                description=get_hdf5_table_description(integral.used_variables, decimal_precision),
                expectedrows=num_rows
            )

            # Add the integral table metadata, as per the section 3.2 of my PhD thesis
            table.attrs.used_variables_list = np.array(integral.used_variables, dtype=str)
            table.attrs.used_variables_num = len(integral.used_variables)

            # OPTIMIZATION: Rationalize the number of integration variables, as per the section 3.3.3 of my PhD thesis
            row = table.row
            for m, t, v, n in integral.iterate_over_used_variables(max_mode=max_mode):
                # Store the values of `used_variables`
                d = locals()
                for var in integral.used_variables:
                    row[var] = d[var]

                # Store the integral and error data
                cache_key = integral.cache_key(m, t, v, n, max_mode=max_mode)
                data = integral_tables[integral.id][cache_key]
                for k, v in data.items():
                    row[k] = v

                row.append()

            # Flush the table manually, or the last data chunk won't be filled with correct values
            table.flush()

            # Create a completely sorted index (CSI) on all `used_variables` columns
            for var in integral.used_variables:
                table.cols._f_col(var).create_csindex()

            table.close()
'''
